internal/shared/xmlbase.go

```diff
@@ -0,0 +1,257 @@
+package shared
+
+import (
+	"bytes"
+	"fmt"
+	"golang.org/x/net/html"
+	"net/url"
+	"strings"
+
+	"github.com/mmcdole/goxpp"
+)
+
+var (
+	// HTML attributes which contain URIs
+	// https://pythonhosted.org/feedparser/resolving-relative-links.html
+	// To catch every possible URI attribute is non-trivial:
+	// https://stackoverflow.com/questions/2725156/complete-list-of-html-tag-attributes-which-have-a-url-value
+	htmlURIAttrs = map[string]bool{
+		"action":     true,
+		"background": true,
+		"cite":       true,
+		"codebase":   true,
+		"data":       true,
+		"href":       true,
+		"poster":     true,
+		"profile":    true,
+		"scheme":     true,
+		"src":        true,
+		"uri":        true,
+		"usemap":     true,
+	}
+)
+
+type urlStack []*url.URL
+
+func (s *urlStack) push(u *url.URL) {
+	*s = append([]*url.URL{u}, *s...)
+}
+
+func (s *urlStack) pop() *url.URL {
+	if s == nil || len(*s) == 0 {
+		return nil
+	}
+	var top *url.URL
+	top, *s = (*s)[0], (*s)[1:]
+	return top
+}
+
+func (s *urlStack) top() *url.URL {
+	if s == nil || len(*s) == 0 {
+		return nil
+	}
+	return (*s)[0]
+}
+
+type XMLBase struct {
+	stack    urlStack
+	URIAttrs map[string]bool
+}
+
+// FindRoot iterates through the tokens of an xml document until
+// it encounters its first StartTag event.  It returns an error
+// if it reaches EndDocument before finding a tag.
+func (b *XMLBase) FindRoot(p *xpp.XMLPullParser) (event xpp.XMLEventType, err error) {
+	for {
+		event, err = b.NextTag(p)
+		if err != nil {
+			return event, err
+		}
+		if event == xpp.StartTag {
+			break
+		}
+
+		if event == xpp.EndDocument {
+			return event, fmt.Errorf("Failed to find root node before document end.")
+		}
+	}
+	return
+}
+
+// XMLBase.NextTag iterates through the tokens until it reaches a StartTag or
+// EndTag It maintains the urlStack upon encountering StartTag and EndTags, so
+// that the top of the stack (accessible through the CurrentBase() and
+// CurrentBaseURL() methods) is the absolute base URI by which relative URIs
+// should be resolved.
+//
+// NextTag is similar to goxpp's NextTag method except it wont throw an error
+// if the next immediate token isnt a Start/EndTag.  Instead, it will continue
+// to consume tokens until it hits a Start/EndTag or EndDocument.
+func (b *XMLBase) NextTag(p *xpp.XMLPullParser) (event xpp.XMLEventType, err error) {
+	for {
+
+		if p.Event == xpp.EndTag {
+			// Pop xml:base after each end tag
+			b.pop()
```

Comment:
Shouldn't this `b.pop()` be in the existing:

```go
if event == xpp.EndTag {
```

Reply:
Not sure what else we could do...

I suppose `DecodeElement` could be proxied through xmlbase as well, but I'm not sure if that is a ton better.

Reply:
I'd lean toward leaving things the way they are now because 1) the `push` and `pop` are relegated to a single function, and 2) you are still free to call xpp directly without needing to remember to use a xmlbase proxy.

But if things prove to be fragile in the future, then confining anything that mutates the parser state to `xmlbase.go` sounds like a good policy.